crates/ra_ide/src/references/rename.rs

```diff
@@ -124,6 +127,116 @@ fn rename_mod(
+        Some(ast::TypeRef::ReferenceType(rt)) => rt.mut_token().is_some(),
+        _ => return None, // not renaming other types
+    };
+
+    let RangeInfo { range, info: refs } = find_all_refs(db, position, None)?;
+
+    let param_range = first_param.syntax().text_range();
+    let (param_ref, usages): (Vec<Reference>, Vec<Reference>) = refs
+        .into_iter()
+        .partition(|reference| param_range.intersect(reference.file_range.range).is_some());
+
+    if param_ref.is_empty() {
+        return None;
+    }
+
+    let mut edits = usages
+        .into_iter()
+        .map(|reference| source_edit_from_reference(reference, "self"))
+        .collect::<Vec<_>>();
+
+    edits.push(SourceFileEdit {
+        file_id: position.file_id,
+        edit: TextEdit::replace(
+            param_range,
+            String::from(if mutable { "&mut self" } else { "&self" }),
+        ),
+    });
+
+    Some(RangeInfo::new(range, SourceChange::source_file_edits("rename", edits)))
+}
+
+fn source_edit_from_self(
+    syn: &SyntaxNode,
+    position: FilePosition,
+    new_name: &str,
+) -> Option<SourceFileEdit> {
+    fn target_type_name(impl_def: &ast::ImplDef) -> Option<String> {
+        if let Some(ast::TypeRef::PathType(p)) = impl_def.target_type() {
+            return Some(p.path()?.segment()?.name_ref()?.text().to_string());
+        }
+        None
+    }
+
+    let token = syn.token_at_offset(position.offset).find(|t| t.kind() == SyntaxKind::SELF_KW)?;
+    let mut replacement_text = String::from(new_name);
+    let range = if let Some(self_param) = ast::SelfParam::cast(token.parent()) {
+        let impl_def = find_node_at_offset::<ast::ImplDef>(syn, position.offset)?;
+        let type_name = target_type_name(&impl_def)?;
+        replacement_text.push_str(": ");
+        replacement_text.push_str(self_param.mut_token().map_or("&", |_| "&mut "));
+        replacement_text.push_str(type_name.as_str());
+        self_param.syntax().text_range()
+    } else {
+        token.text_range()
+    };
+    Some(SourceFileEdit {
+        file_id: position.file_id,
+        edit: TextEdit::replace(range, replacement_text),
+    })
+}
+
+fn rename_self_to_param(
+    db: &RootDatabase,
+    position: FilePosition,
+    new_name: &str,
+) -> Option<RangeInfo<SourceChange>> {
+    use std::convert::TryInto;
+
+    let sema = Semantics::new(db);
+    let source_file = sema.parse(position.file_id);
+    let syn = source_file.syntax();
+
+    let token = syn.token_at_offset(position.offset).find(|t| t.kind() == SyntaxKind::SELF_KW)?;
+    let range = ast::SelfParam::cast(token.parent())
+        .map_or(token.text_range(), |p| p.syntax().text_range());
+
+    let text = db.file_text(position.file_id);
+    let fn_def = find_node_at_offset::<ast::FnDef>(syn, position.offset)?;
+    let search_range = fn_def.syntax().text_range();
+
+    let mut edits: Vec<SourceFileEdit> = vec![];
+
+    for (idx, _) in text.match_indices("self") {
```

Comment:
Heh, I thnink ideally we should just make semantic goto def and find references work for `self` (they don't at the moment), but this seems to be a surprisngly good enough approach.

Reply:
Do you mean adding another item in `Definition` enum? Any suggestion for the name?
I'm still struggling with understanding what it is exactly :)
I'll try that, might be a bit harder, but should be cleaner to keep that functionality together.

Reply:
Would that also mean adding new `ReferenceKind`?

Reply:
>Do you mean adding another item in Definition enum?

I think `local` already can refer to self, it's just that we don't handle it in goto def and find reference. 

>I'll try that, might be a bit harder, but should be cleaner to keep that functionality together.

I think it makes sense to 

* *first* merge this mostly as is, 
* *second*, fix goto definition and find renferences to work for `self`
* *third*, maybe optional, do the #4290 refactorings
* finally, re-reail the current logic onto semantic self. 

>I'm still struggling with understanding what it is exactly :)

Good question. So in the compiler we don't actually have a generic, unified concept of symbol or reference. That's because the set of types of things a local variable can resolve to is different from what a path can resolve to, and we try to be well-typed.

OTOH in an IDE, when you do, say rename, you expect it to work exactly the same for locals, lifetimes, modules etc. So we define an IDE-specific `Definition` enum as "something, which can be refered to by name".

Reply:
Thanks for explanation and suggestions, will follow this steps.